### PR TITLE
Fix DynamoDB helpers to return deserialized Python types

### LINKED ISSUE #244: Fix DynamoDB deserialization bug in get_item_from_table
## 🐛 Bug Description

The `get_item_from_table` method in `lib/aws/dynamodb.py` returns DynamoDB's native serialized format (e.g., `{"service": {"S": "value"}, "cursor": {"N": "123"}}`), but multiple callers expect plain Python types (e.g., `{"service": "value", "cursor": 123}`). This causes runtime failures when passing results to Pydantic model constructors.

**Root Cause:**
`boto3.client.get_item()` returns items in DynamoDB's native format with type descriptors (`{"S": ...}`, `{"N": ...}`, etc.), but the code passes these directly to Pydantic models that expect plain Python types.

### Primary Issue
- `lib/aws/dynamodb.py:74-81` - `get_item_from_table` method returns serialized format
- `lib/aws/dynamodb.py:92-99` - `get_all_items_from_table` has the same issue

### Broken Call Sites (Will Fail at Runtime)
1. `services/sync/stream/cursor.py:29-34` - Passes result to `FirehoseSubscriptionStateCursorModel(**result)`
2. `services/sync/stream/deprecated/export_data.py:765-770` - Same issue
3. `deprecated/distributed_job_coordination/lib/dynamodb_utils.py:58-59` - Passes to `JobState(**job_state_dict)`
4. `deprecated/distributed_job_coordination/lib/dynamodb_utils.py:112-113` - Passes to `TaskState(**task_state_dict)`
5. `deprecated/distributed_job_coordination/lib/dynamodb_utils.py:120-124` - Uses `get_all_items_from_table` and passes to Pydantic models

### Working Call Sites (No Changes Needed)
- `lib/aws/dynamodb.py:66` - Used in `update_item_in_table` (works because `insert_item_into_table` handles both formats)
- `lib/aws/dynamodb.py:86` - Used in `verify_item_exists` (only checks existence, doesn't use data)

### Current Production Usage
- The production code in `services/sync/stream/firehose.py` uses `load_cursor_state_s3` (S3 version), not the DynamoDB version, so this bug may not be hit in production yet.
- However, the DynamoDB functions exist and will fail if/when used.

### Severity
- **High**: All 5 broken call sites will fail at runtime with Pydantic validation errors
- **Medium**: The bug exists in both `get_item_from_table` and `get_all_items_from_table`

## 💡 Proposed Solution

Add automatic deserialization in `get_item_from_table` and `get_all_items_from_table` methods using `boto3.dynamodb.types.TypeDeserializer`.

### Implementation Details

1. **Add TypeDeserializer import and instance:**
   ```python
   from boto3.dynamodb.types import TypeDeserializer, TypeSerializer
   deserializer = TypeDeserializer()
   ```

2. **Add helper method for deserialization:**
   ```python
   def _deserialize_item(self, item: Optional[dict]) -> Optional[dict]:
       """Deserialize a DynamoDB item to plain Python types."""
       if item is None:
           return None
       return {k: deserializer.deserialize(v) for k, v in item.items()}
   ```

3. **Update `get_item_from_table`:**
   - Deserialize the item before returning
   - Update docstring to clarify it returns deserialized data

4. **Update `get_all_items_from_table`:**
   - Deserialize all items before returning
   - Update docstring to clarify it returns deserialized data

### Why This Approach?

- ✅ **Fixes all broken call sites automatically** - No changes needed in 5 different files
- ✅ **Backward compatible** - `update_item_in_table` will continue to work because `insert_item_into_table` already handles both serialized and deserialized formats (checks `is_serialized()`)
- ✅ **Consistent API** - All methods return deserialized data
- ✅ **Also fixes `get_all_items_from_table`** - Which has the same issue

## ✅ Expected Outcomes

After this fix:
- All 5 broken call sites will work correctly with Pydantic models
- `update_item_in_table` will continue to work (no breaking changes)
- `verify_item_exists` will continue to work (no changes needed)
- API becomes consistent - all get methods return deserialized data

## 🧪 Testing Recommendations

1. Test `load_cursor_state_dynamodb` with actual DynamoDB data
2. Test `update_item_in_table` to ensure it still works correctly
3. Test deprecated code paths if they're still in use
4. Add unit tests for the deserialization logic

## 📝 Additional Notes

- This issue was identified by CodeRabbit AI in a code review
- The same pattern exists in `get_all_items_from_table` which should be fixed as well
- The codebase already has `TypeSerializer` imported and used, so adding `TypeDeserializer` follows existing patterns

### PR DESCRIPTION
**Pull request by mark-torres10**
## Summary
- Fix `lib/aws/dynamodb.py` to deserialize DynamoDB `Item`/`Items` into plain Python types (including numeric normalization) for `get_item_from_table` and `get_all_items_from_table`.
- Update a few call sites that were previously indexing into DynamoDB's typed format (e.g. `item["field"]["S"]`).
- Add unit tests covering deserialization behavior.
 
## Issue
Fixes #244
 
Reference: https://github.com/METResearchGroup/bluesky-research/issues/244
 
## Test plan
- `pytest lib/tests/test_dynamodb_deserialization.py`
- `ruff check`

---
<a href="https://cursor.com/background-agent?bcId=bc-39582393-c401-4a3e-95e6-315dda0a1861"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-cursor-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-cursor-light.svg">Open in Cursor</picture></a>&nbsp;<a href="https://cursor.com/agents?id=bc-39582393-c401-4a3e-95e6-315dda0a1861"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/open-in-web-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/open-in-web-light.svg">Open in Web</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added query helpers (by service, inference type, custom index) and an index existence check.
  * Retrieval helpers now return deserialized, normalized native Python types (ints/floats, lists, dicts, nulls).

* **Bug Fixes**
  * Handle empty-result paths gracefully.
  * Consistent timestamp extraction/handling across services.

* **Tests**
  * Added unit tests for deserialization and related retrieval behaviors.

* **Chores**
  * Updated project packaging configuration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: lib/aws/dynamodb.py [hunk 1]
```diff
@@ -90,10 +132,11 @@ def verify_item_exists(self, key: dict, table_name: str) -> bool:
 
     @retry_on_aws_rate_limit
     def get_all_items_from_table(self, table_name: str) -> list[dict]:
-        """Gets all items from a table."""
+        """Gets all items from a table (deserialized)."""
         try:
             response = self.client.scan(TableName=table_name)
-            return response.get("Items")
+            items = response.get("Items", []) or []
```
Comment (mark-torres10) [inline lib/aws/dynamodb.py]: Why do we have the "or []" part at the end? It looks like the .get already handles default behavior.

File: lib/aws/dynamodb.py [hunk 2]
```diff
@@ -37,6 +41,41 @@ def is_serialized(self, item: dict) -> bool:
...
+        if isinstance(value, list):
+            return [self._normalize_deserialized_value(v) for v in value]
+        if isinstance(value, dict):
+            return {k: self._normalize_deserialized_value(v) for k, v in value.items()}
+        if isinstance(value, set):
+            return {self._normalize_deserialized_value(v) for v in value}
+        return value
+
+    def _deserialize_item(self, item: dict) -> dict:
+        """Deserialize a DynamoDB-typed item into plain Python types."""
```
Comment (mark-torres10) [inline lib/aws/dynamodb.py]: Could you put an input/output example in the docstring so it's easier to know how this will look?

File: lib/aws/dynamodb.py [hunk 3]
```diff
@@ -37,6 +41,41 @@ def is_serialized(self, item: dict) -> bool:
...
+        deserialized = {k: deserializer.deserialize(v) for k, v in item.items()}
+        normalized = self._normalize_deserialized_value(deserialized)
+        if not isinstance(normalized, dict):
```
Comment (mark-torres10) [inline lib/aws/dynamodb.py]: Is this here for defensive reasons or is there a specific error case that we're defending against here?
